### PR TITLE
Basic Performance Framework Code

### PR DESCRIPTION
**Pull request by pavan187**
**What type of PR is this?**

/kind test

**What this PR does / why we need it**:
This PR has Basic Performance Framework code, which tests the deployment of kubeedge(edgecore) in K8s POD and verify the EdgeNode are up and Running, then do the application deployment on EdgeNodes.

**Which issue(s) this PR fixes**:
Partial Fixes # PR-308

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by m1093782566** [comment]
@pavan187 I think I have found a bug of CI-BOT :) People are assigned multiple times.

**Comment by pavan187** [comment]
> @pavan187 I think I have found a bug of CI-BOT :) People are assigned multiple times.

@m1093782566 , i also just saw, will fix it :-)

File: tests/performance/loadtest/loadtest_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
+/*
```
Comment (edisonxiang) [inline tests/performance/loadtest/loadtest_suite_test.go]: please remove `tests/performance/loadtest/loadtest.test`

File: tests/performance/loadtest/loadtest_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+				if status == 404{
+					count++
+				}
+			}
+			return count
+		}, "60s", "4s").Should(Equal(ctx.Cfg.NumOfNodes), "EdgeNode deleton is unsuccessfull !!")
+	})
+
+	RunSpecs(t, "kubeedge Performace Load test Suite")
+}
```
Comment (edisonxiang) [inline tests/performance/loadtest/loadtest_suite_test.go]: Add newline

File: tests/performance/loadtest/loadtest_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+					podlist, err = utils.GetPods(ctx.Cfg.ApiServer + AppHandler, label)
+					Expect(err).To(BeNil())
+					break
+				}
+			}
+			utils.CheckPodRunningState(ctx.Cfg.ApiServer + AppHandler, podlist)
+		})
+
+	})
+})
```
Comment (edisonxiang) [inline tests/performance/loadtest/loadtest_test.go]: ditto

File: tests/performance/scripts/build_upload_image.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    docker rmi -f $(docker images -q ${imageRepo}/edgecore:latest)
+fi
+
+cd ${SRC_DIR}
+
+docker build -t ${imageRepo}/edgecore:latest -f ${SRC_DIR}/build/edge/Dockerfile .
+
+docker push ${imageRepo}/edgecore:latest
+
+echo "edgecore image successully built and pushed to repository !!"
```
Comment (edisonxiang) [inline tests/performance/scripts/build_upload_image.sh]: ditto

**Comment by edisonxiang** [comment]
@pavan187 Well done. Just minor comments.

**Comment by pavan187** [comment]
> @pavan187 I think I have found a bug of CI-BOT :) People are assigned multiple times.

@m1093782566 , This is not a CI-BOT issue, @srivatsav123 mistakenly use CI-BOT code in his **winner BOT" which will handle the all GIT events handled by **kubeedge-bot**. so kubeedge-bot and winner BOT both were adding reviewers.  Now @srivatsav123 removed the CI_BOT code in his Winner BOT, now it works fine.

**Comment by m1093782566** [comment]
Fine @pavan187

**Comment by edisonxiang** [comment]
@pavan187 We need to update this PR base on the deployment proposal
https://github.com/kubeedge/kubeedge/blob/eba055b22a5dc30705798bacb0afefaf1249c2e2/docs/proposals/perf.md#performance-test-deployment

**Comment by pavan187** [comment]
> @pavan187 We need to update this PR base on the deployment proposal
> https://github.com/kubeedge/kubeedge/blob/eba055b22a5dc30705798bacb0afefaf1249c2e2/docs/proposals/perf.md#performance-test-deployment

Yes, will update the PR in sometime.

**Comment by m1093782566** [comment]
@pavan187 

Any update?

File: tests/e2e/utils/common.go [hunk 5]
```diff
@@ -26,59 +26,133 @@ import (
...
+		Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"app": "edgecore"}},
+		Template: v1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{"app": "edgecore"},
+			},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:            "edgecore",
+						Image:           "pavan187/edgecore:latest",
```
Comment (sids-b) [inline tests/e2e/utils/common.go]: can this image be uploaded to kubeedge docker registry ? 
cc @m1093782566

File: tests/e2e/utils/common.go [hunk 6]
```diff
@@ -26,59 +26,133 @@ import (
...
+		TypeMeta: metav1.TypeMeta{APIVersion: "apps/v1", Kind: "Deployment"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Labels:    map[string]string{"app": "edgecore"},
+			Namespace: namespace,
+		},
+		Spec: *depObj,
+	}
 	return &deployment
 }
```
Comment (sids-b) [inline tests/e2e/utils/common.go]: please add a blank line after this

File: tests/e2e/utils/common.go [hunk 7]
```diff
@@ -97,6 +171,15 @@ func GetDeployments(list *apps.DeploymentList, getDeploymentApi string) error {
 	return nil
 
 }
```
Comment (sids-b) [inline tests/e2e/utils/common.go]: please add a blank line after this

File: tests/e2e/utils/node.go [hunk 8]
```diff
@@ -124,3 +138,56 @@ func CheckNodeDeleteStatus(ctx *TestContext, nodehandler, nodename string) int {
 	defer resp.Body.Close()
 	return resp.StatusCode
 }
+
+//HandleConfigmap function to create configmaps for respective edgenodes
+func HandleConfigmap(ctx *TestContext, configName chan error, operation, confighandler string) {
+	var req *http.Request
+	curpath := getpwd()
+	file := path.Join(curpath, "../../../build/edge/kubernetes/03-configmap-edgenodeconf.yaml")
```
Comment (sids-b) [inline tests/e2e/utils/node.go]: please use path from $GOPATH
Applicable to all similar places

File: tests/e2e/utils/node.go [hunk 9]
```diff
@@ -124,3 +138,56 @@ func CheckNodeDeleteStatus(ctx *TestContext, nodehandler, nodename string) int {
...
+		}
+		InfoV6("%s %s %v in %v", req.Method, req.URL, resp.Status, time.Now().Sub(t))
+		defer resp.Body.Close()
+		if operation == http.MethodPost {
+			Expect(resp.StatusCode).Should(Equal(http.StatusCreated))
+		} else {
+			Expect(resp.StatusCode).Should(Equal(http.StatusOK))
+		}
+		configName <- nil
+
```
Comment (sids-b) [inline tests/e2e/utils/node.go]: please remove blank line

File: tests/e2e/utils/node.go [hunk 10]
```diff
@@ -124,3 +138,56 @@ func CheckNodeDeleteStatus(ctx *TestContext, nodehandler, nodename string) int {
...
+//GetConfigmap function to get configmaps for respective edgenodes
+func GetConfigmap(apiConfigMap string) int {
+	err, resp := SendHttpRequest(http.MethodGet, apiConfigMap)
+	if err != nil {
+		Failf("Sending SenHttpRequest failed: %v", err)
+		return -1
+	}
+	defer resp.Body.Close()
+	return resp.StatusCode
+
```
Comment (sids-b) [inline tests/e2e/utils/node.go]: please remove blank line

File: tests/performance/loadtest/loadtest_suite_test.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+			for edgenodeName,_ := range nodeInfo{
+				status := utils.CheckNodeReadyStatus(ctx, NodeHandler, edgenodeName)
+				utils.Info("Node Name: %v, Node Status: %v", edgenodeName, status)
+				if status == "Running"{
+					count++
+				}
+			}
+			return count
+		}, "60s", "4s").Should(Equal(ctx.Cfg.NumOfNodes), "Nodes register to the k8s master is unsuccessfull !!")
+
```
Comment (sids-b) [inline tests/performance/loadtest/loadtest_suite_test.go]: please remove blank line

File: tests/performance/loadtest/loadtest_test.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+				if deployment.Name == UID {
+					label := nodeName
+					podlist, err = utils.GetPods(ctx.Cfg.ApiServer + AppHandler, label)
+					Expect(err).To(BeNil())
+					break
+				}
+			}
+			utils.CheckPodRunningState(ctx.Cfg.ApiServer + AppHandler, podlist)
+		})
+
```
Comment (sids-b) [inline tests/performance/loadtest/loadtest_test.go]: please remove blank line

File: tests/e2e/utils/common.go [hunk 13]
```diff
@@ -136,16 +219,16 @@ func HandlePod(operation string, apiserver string, UID string, ImageUrl, nodesel
...
 
 //HandleDeployment to handle app deployment/delete deployment.
-func HandleDeployment(operation, apiserver, UID, ImageUrl, nodeselector string, replica int) bool {
+func HandleDeployment(IsEdgeCore bool, operation, apiserver, UID, ImageUrl, nodeselector, configmapname string, replica int) bool {
```
Comment (sids-b) [inline tests/e2e/utils/common.go]: nit: IsEdgeCore -> isEdgeCore
Applicable to all similar places

File: tests/performance/loadtest/loadtest_suite_test.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+//stores deployments info
+var deployments []string
+//chan for goroutines
+var chconfigmapRet = make(chan error)
+
+//Function to run the Ginkgo Test
+func TestEdgecoreK8sDeployment(t *testing.T) {
+	RegisterFailHandler(Fail)
+	var _ = BeforeSuite(func() {
+		utils.InfoV6("Kubeedge deployment Load test Bigin !!")
```
Comment (rohitsardesai83) [inline tests/performance/loadtest/loadtest_suite_test.go]: Bigin -> begin

File: tests/performance/scripts/update_configmap.sh [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    if [ ! -f ${EDGE_PATH} ]; then
+        echo "There is no 03-configmap-edgenodeconf.yaml!"
+        exit 1
+    fi
+    echo "file found !!!!!!!!!!!!!"
+    sed -i "s|namespace: .*|namespace: default|g" ${EDGE_PATH}
+    sed -i "s|name: .*|name: ${configmapName}|g" ${EDGE_PATH}
+    sed -i "s|node-id: .*|node-id: ${nodename}|g" ${EDGE_PATH}
+    sed -i "s|hostname-override: .*|hostname-override: ${nodename}|g" ${EDGE_PATH}
+    sed -i "s|url: .*|url: ${websockerUrl}/e632aba927ea4ac2b575ec1603d56f10/${nodename}/events|g" ${EDGE_PATH}
```
Comment (rohitsardesai83) [inline tests/performance/scripts/update_configmap.sh]: e632aba927ea4ac2b575ec1603d56f10  -> hardcoded ?

File: tests/performance/loadtest/loadtest_test.go [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+			for _, deployment := range deploymentList.Items {
+				if deployment.Name == UID {
+					//label := nodeName
+					time.Sleep(1 *time.Second)
+					podlist, err = utils.GetPods(ctx.Cfg.ApiServer2 + AppHandler, "")
+					Expect(err).To(BeNil())
+					break
+				}
+			}
+			fmt.Println("*****************************", len(podlist.Items))
```
Comment (rohitsardesai50) [inline tests/performance/loadtest/loadtest_test.go]: please cleanup logs.

File: tests/e2e/scripts/fast_test.sh [hunk 17]
```diff
@@ -32,7 +32,7 @@ cat >config.json<<END
 {
         "edgedEndpoint": "http://127.0.0.1:10255",
         "image_url": ["nginx", "hello-world"],
-        "apiserver":"http://$MASTER_IP:12418"
+        "K8smaster1":"http://$MASTER_IP:12418"
```
Comment (sids-b) [inline tests/e2e/scripts/fast_test.sh]: please dont suffix with number.

**Comment by m1093782566** [comment]
I am going to merge this PR, we can address the outstanding review comments in another PR.

**Pull request merged**